client/asset/dcr/dcr.go

```diff
@@ -0,0 +1,1165 @@
+	if err != nil {
+		return nil, err
+	}
+	txOut, err := dcr.node.GetTxOut(txHash, vout, true)
+	if err != nil {
+		return nil, fmt.Errorf("error finding unspent output %s:%d: %v", txHash, vout, err)
+	}
+	if txOut == nil {
+		return nil, fmt.Errorf("tx output not found")
+	}
+	pkScript, err := hex.DecodeString(txOut.ScriptPubKey.Hex)
+	if err != nil {
+		return nil, err
+	}
+	tree := wire.TxTreeRegular
+	if dexdcr.ParseScriptType(dexdcr.CurrentScriptVersion, pkScript, coin.Redeem()).IsStake() {
+		tree = wire.TxTreeStake
+	}
+	return newOutput(dcr.node, txHash, vout, coin.Value(), tree, coin.Redeem()), nil
+}
+
+// sendWithReturn sends the unsigned transaction with an added output (unless
+// dust) for the change.
+func (dcr *ExchangeWallet) sendWithReturn(baseTx *wire.MsgTx,
+	addr dcrutil.Address, totalIn, totalOut uint64) (*wire.MsgTx, error) {
+
+	msgTx, signed, err := dcr.node.SignRawTransaction(baseTx)
+	if err != nil {
+		return nil, fmt.Errorf("signing error: %v", err)
+	}
+	if !signed {
+		return nil, fmt.Errorf("incomplete raw tx signature")
+	}
+	size := msgTx.SerializeSize()
+	minFee := dcr.nfo.FeeRate * uint64(size)
+	remaining := totalIn - totalOut
+	if minFee > remaining {
+		return nil, fmt.Errorf("not enough funds to cover minimum fee rate. %d < %d",
+			totalIn, minFee+totalOut)
+	}
+	changeScript, err := txscript.PayToAddrScript(addr)
+	if err != nil {
+		return nil, fmt.Errorf("error creating change script: %v", err)
+	}
+	changeIdx := len(baseTx.TxOut)
+	changeOutput := wire.NewTxOut(1, changeScript)
+	changeOutput.Value = int64(remaining - minFee)
+	isDust := dexdcr.IsDust(changeOutput, dcr.nfo.FeeRate)
+	sigCycles := 0
+	if !isDust {
+		// Add the change output with a recalculated fees
+		size = size + dexdcr.P2PKHOutputSize
+		fee := dcr.nfo.FeeRate * uint64(size)
+		changeOutput.Value = int64(remaining - fee)
+		baseTx.AddTxOut(changeOutput)
+		// Find the best fee rate by closing in on it in a loop.
+		tried := map[uint64]uint8{}
+		for {
+			sigCycles++
+			msgTx, signed, err = dcr.node.SignRawTransaction(baseTx)
+			if err != nil {
+				return nil, fmt.Errorf("signing error: %v", err)
+			}
+			if !signed {
+				return nil, fmt.Errorf("incomplete raw tx signature")
+			}
+			size := msgTx.SerializeSize()
+			reqFee := dcr.nfo.FeeRate * uint64(size)
+			if reqFee > remaining {
+				// I can't imagine a scenario where this condition would be true, but
+				// I'd hate to be wrong.
+				dcr.log.Errorf("reached the impossible place. in = %d, out = %d, reqFee = %d, lastFee = %d",
+					totalIn, totalOut, reqFee, fee)
+				return nil, fmt.Errorf("change error")
+			}
+			if fee == reqFee || (reqFee < fee && tried[reqFee] == 1) {
+				// If we've already tried it, we're on our way back up, so this fee is
+				// likely as good as it gets.
+				break
+			}
+			// We must have some room for improvement
+			tried[fee] = 1
+			fee = reqFee
+			changeOutput.Value = int64(remaining - fee)
+			continue
```

Comment:
You reworked the loop in the BTC wallet PR.  Anything to port over from that to here?  I didn't compare the two, but this looks like the reworked version.

Reply:
Good call. I forgot when I had translated, and there were a couple of changes I had missed.